cpp/src/spatial/lonlat_to_coordinate.cu

```diff
@@ -0,0 +1,113 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include <type_traits>
+#include <utility>
+#include <thrust/for_each.h>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cuspatial/error.hpp>
+
+namespace {
+
+using pair_of_columns = std::pair<std::unique_ptr<cudf::column>, std::unique_ptr<cudf::column>>;
+
+struct ll2coord_functor
+{
+    template <typename T, typename... Args>
+    std::enable_if_t<not std::is_floating_point<T>::value, pair_of_columns>
+    operator()(Args&&...)
+    {
+        CUSPATIAL_FAIL("Non-floating point operation is not supported");
+    }
+
+    template <typename T>
+    std::enable_if_t<std::is_floating_point<T>::value, pair_of_columns>
+    operator()(double origin_lon,
+               double origin_lat,
+               cudf::column_view const& input_lon,
+               cudf::column_view const& input_lat,
+               rmm::mr::device_memory_resource *mr,
+               cudaStream_t stream)
+    {
+        auto size = input_lon.size();
+        auto tid = cudf::experimental::type_to_id<T>();
+        auto type = cudf::data_type{ tid };
+        auto d_input_lon = cudf::column_device_view::create(input_lon);
+        auto d_input_lat = cudf::column_device_view::create(input_lat);
+        auto output_lon = cudf::make_fixed_width_column(type, size, cudf::mask_state::UNALLOCATED, stream, mr);
+        auto output_lat = cudf::make_fixed_width_column(type, size, cudf::mask_state::UNALLOCATED, stream, mr);
+        auto d_output_lon = cudf::mutable_column_device_view::create(output_lon->mutable_view());
+        auto d_output_lat = cudf::mutable_column_device_view::create(output_lat->mutable_view());
+
+        auto iter = thrust::make_counting_iterator<cudf::size_type>(0);
+
+        thrust::for_each(iter,
+                         iter + size,
+                         [origin_lon=origin_lon,    origin_lat=origin_lat,
+                              in_lon=*d_input_lon,      in_lat=*d_input_lat,
+                             out_lon=*d_output_lon,    out_lat=*d_output_lat]
+                         __device__(cudf::size_type idx) mutable
+                         {
+                            out_lon.element<T>(idx) = (origin_lon - in_lon.element<T>(idx))
+                                * 40000.0
+                                * cos((origin_lat + in_lon.element<T>(idx)) * M_PI / 360)
+                                / 360;
+
+                            out_lat.element<T>(idx) = (origin_lat - in_lat.element<T>(idx))
+                                * 40000.0
+                                / 360;
+                         });
```

Comment:
A question for @harrism  - isn't this the perfect situation to use a `thrust::transform` iterator, instead of a `for_each`, since the operation is identical and independent for every element of the input vectors?

Reply:
I believe using `transform` in this scenario would require two kernel invocations (one for lon, one for lat) and reading the latitude column twice (since it is used in the longitude calculation). Using `for_each` allows us to launch a single kernel and only read each input column once.

It would be nice to have a transform function which took a tuple of columns and returned a tuple of columns, accepted a tuple of Ts and returned a tuple of Ts, but that doesn't exist (until we write one).

Reply:
I think `thrust::zip_iterator` can only be used as an input iterator, not an output iterator. If we had something like `make_zip_output_iterator()`, we could use transform to read the two input columns and return a `thrust::tuple()` of the two output values.

Reply:
> I think `thrust::zip_iterator` can only be used as an input iterator, not an output iterator. If we had something like `make_zip_output_iterator()`, we could use transform to read the two input columns and return a `thrust::tuple` of the two output values.

Zip_iterator worked for me as both inputs and outputs.

Reply:
@zhangjianting are there docs or an example for this?

Reply:
> @zhangjianting are there docs or an example for this?

cpp/tests/spatial/coordinate_transform_test.cu

Reply:
Thanks for the insight, everyone. I was thinking about this from the wrong perspective, but zip iterators as output iterators is great!